### PR TITLE
[github] Include commit comments in Github backend

### PR DESCRIPTION
**Pull request by harshalmittal4**
Addresses #515.
The changes made:

- `commits_data`  field in pull_request response originally contained the PR commit `hash`, now it contains `commit_comments` as well. 

- `commit_comments`  has two information : comment body and user_data (for the comment author).

Currently, the PR is not ready to be merged.

It is running into error `perceval.errors.ArchiveError: data storage error; cause: duplicated entry`, 
(the error can be reproduced using command  `perceval github Cloud-CV Origami --json-line --category pull_request --sleep-for-rate -t XXXX>> testing.json
` , for any other other repo also, same error is popping up).
It would really be helpful if someone could have a look where am I missing!

Also, I wished to understand if any information needs to be removed or added in `commits_data`.

I will add the tests to the PR once this is done, thanks a lot :)
@valeriocos @aswanipranjal @jgbarah

### CODE REVIEW

**Comment by harshalmittal4** [comment]
Also, is it okay to include the `commit_comments` field inside `commits_data`, or shall it be added seperately to the pull_request response, please see

**Comment by harshalmittal4** [comment]
@valeriocos, I need a little help to figure out what the problem is, please just take a look whenever you have time.

**Comment by Polaris000** [comment]
I get the same errors with mine, @harshalmittal4. Not completely sure why.

**Comment by valeriocos** [comment]
I'll have a look at it ASAP, thanks for the reminder :)

**Comment by valeriocos** [comment]
When perceval fetches data from remote sources (e.g., GitHub API), it stores the requests and the corresponding responses in a database/archive (see details [here](https://github.com/chaoss/grimoirelab-perceval/blob/master/perceval/client.py#L160)) in order to reuse this data for other executions. In case the combination of `url, payload, headers, response` already exists in the archive, the following error is thrown: `perceval.errors.ArchiveError: data storage error; cause: duplicated entry` (ideally the same request shouldn't be issued more than once).

In the case of commits in PRs, it may happen that the same commit is present in more pull requests, for instance:
- https://github.com/chaoss/grimoirelab/pull/7/commits
- https://github.com/chaoss/grimoirelab/pull/8/commits

A possible solution to avoid the archive error is to use an internal cache (as similarly done for github [users](https://github.com/chaoss/grimoirelab-perceval/blob/master/perceval/backends/core/github.py#L677)).

@Polaris000  @harshalmittal4 at the moment this is the only solution that comes to my mind. However for repos with many commits in PRs, this solution may not be good. 

What do you think ?

**Comment by harshalmittal4** [comment]
Hey @valeriocos, the error is resolved by using an internal cache, other than that, can you just take a look at the PR structure, I mean this :

The changes made:
- commits_data field in pull_request response originally contained the PR commit hash, now it contains commit_comments as well.
- commit_comments has two information : comment body and user_data (for the comment author).

I wished to understand if any information needs to be removed or added in commits_data.

Thanks ;)

File: perceval/backends/core/github.py [hunk 1]
```diff
@@ -389,18 +389,40 @@ def __get_pull_requested_reviewers(self, pr_number):
         return requested_reviewers
 
     def __get_pull_commits(self, pr_number):
-        """Get pull request commit hashes"""
+        """Get pull request commit info"""
```
Comment (valeriocos) [inline perceval/backends/core/github.py]: `info` seems to be too generic, you could have something like:
`"""Get pull request commit hashes and comments"""`

**Review by valeriocos — COMMENTED** [review]
@harshalmittal4 the PR looks OK (I left just a minor comment). Please include the caching support and tests.

Thank you!

**Comment by harshalmittal4** [comment]
@valeriocos, the PR isn't ready to be merged, it needs some improvements. will update it soon.

**Comment by valeriocos** [comment]
great @harshalmittal4 let me know when it is ready to review, thanks!

**Comment by harshalmittal4** [comment]
Hey @valeriocos,
I have made the updates, it can be reviewed and let me know for possible improvements  :)

**Comment by valeriocos** [comment]
Thank you @harshalmittal4 , I'll check it later today

**Comment by harshalmittal4** [comment]
Also, the PR is failing due to covergae decrease, so I needed some guidance what can I do about this or if this could be avoided, thanks.

**Comment by valeriocos** [comment]
To check the coverage, you should use the coverage package and execute it as follows:
- coverage run run_tests.py (this will create the file `.coverage`, used to generate the report).
- coverage report -m (this will generate a report and show the lines not covered and percentage per file).

File: perceval/backends/core/github.py [hunk 2]
```diff
@@ -389,18 +389,41 @@ def __get_pull_requested_reviewers(self, pr_number):
...
+
+    def __get_commit_comments(self, commit_sha):
+        """Get pull request commit comments"""
+
+        comments = []
+        group_comments = self.client.pull_commit_comments(commit_sha)
+
+        for raw_comments in group_comments:
+
+            for comment in json.loads(raw_comments):
```
Comment (harshalmittal4) [inline perceval/backends/core/github.py]: Hey @valericos, I checked the coverage report and the lines not covered are line#417 to line#420, and line#425 in the here. I think it is due to the fact that in test_github.py, for the  "zhquan_example/repo", there are no commit comments (the group_comments is an empty list), and so these lines are not executed.
Please let me know what do you think :)

Comment (harshalmittal4) [inline perceval/backends/core/github.py]: Hey @valeriocos , I checked the coverage report and the lines not covered are line#417 to line#420, and line#425 in the here. I think it is due to the fact that in test_github.py, for the  "zhquan_example/repo", there are no commit comments (the group_comments is an empty list), and so these lines are not executed.
Please let me know what do you think, thanks :)

File: tests/test_github.py [hunk 3]
```diff
@@ -575,7 +577,14 @@ def test_fetch_more_pulls(self):
                                    'X-RateLimit-Remaining': '20',
                                    'X-RateLimit-Reset': '5'
                                })
-
+        httpretty.register_uri(httpretty.GET,
+                               GITHUB_REPO_URL + '/commits/53b970ee04bbc435842c14a2cbfdd623faf74a65/comments',
+                               body=[],
```
Comment (harshalmittal4) [inline tests/test_github.py]: Hey @valeriocos, I ain't sure if this is okay, since I am returning an empty list, please have a look.

**Review by valeriocos — COMMENTED** [review]
@harshalmittal4, overall the code looks OK, but the tests need some improvements. 
It seems that writing tests that use httpretty is not easy at the beginning. If so, could you summarize the difficulties you found? 

As suggested to @Polaris000 we could improve the documentation to explain how to write this kind of tests ( maybe a small section in the grimoirelab book @jgbarah ). What do you think ?

I would suggest to take a look at this review: https://github.com/chaoss/grimoirelab-perceval/pull/519 and this commit https://github.com/valeriocos/perceval/commit/01314c90c00da380aa0d4767c6d97b81a367939f (which includes the test data and corresponding modifications in test_github.py for that feature) to better understand how tests work.

I try to summarize how tests that rely on httpretty work using two examples: `test_get_user` and `test_fetch_pulls`.

- [test_get_user](https://github.com/chaoss/grimoirelab-perceval/blob/master/tests/test_github.py#L2574): 
  This test checks that the method `client.user` retrieves the info about a user via the GitHub endpoint `/users/<login>`. Httpretty allows to [mock](https://github.com/chaoss/grimoirelab-perceval/blob/master/tests/test_github.py#L2589) http requests by defining the request url and the output of the response (`GITHUB_USER_URL` and `body=login` in `httpretty.register_uri`), which is available [here](https://github.com/chaoss/grimoirelab-perceval/blob/master/tests/data/github/github_login) . Thus, two important considerations have to be done: 1) the data returned when executing the test `test_get_user` will be always the same, 2) the data returned can be whatever and may not exist in the real API, if you take a look at the content of the github_login file, you will see that `zhquan_example` doesn't exist on GitHub.

Things get a bit more complicated when a test requires to call more API endpoints, this is the case of `test_fetch_pulls`, which mimics the behavior of the backend when fetching pull requests. To make httpretty works, all endpoint calls made by the backend should be mocked, otherwise a call will be issued against the real API, and since the repo `zhquan_example/repo` and the corresponding data don't exist, some HTTP errors will pop up and make the test(s) stuck (this is specially true for user info, if you have a look at the test data you will see that most user info is the one of `zhquan_example` (the mocked user)). 

- [test_fetch_pulls](https://github.com/chaoss/grimoirelab-perceval/blob/master/tests/test_github.py#L255):
  This test mocks the execution of Perceval when fetching a pull request. As you can see, several URIs are mocked ([lines](https://github.com/chaoss/grimoirelab-perceval/blob/master/tests/test_github.py#L268) 268 - 340). Thus, in case the backend is fetching data from a new endpoint (reviews, commit comments, etc), a new URI must be added to mock the corresponding call and mocked data must be provided. It is worth noting that some calls may rely on values included in previous responses, for instance the method `__get_pull_review_comments` reads the pull request [number](https://github.com/chaoss/grimoirelab-perceval/blob/master/perceval/backends/core/github.py#L288) from the pull request obtained. In this case, the attribute `number` in the [mocked data](https://github.com/chaoss/grimoirelab-perceval/blob/master/tests/data/github/github_request_pull_request_1#L336) must be modified to make the tests work. 

If what written above is clear and to answer your question [here](https://github.com/chaoss/grimoirelab-perceval/pull/518/commits/e89f85d058d8f3630eb3661ce0687aebeecb0b4e#r279197338), you should add mocked data for commit comments and make sure that the values embedded in it are consistent with the information needed by the related API calls.

**Comment by harshalmittal4** [comment]
Thanks a lot @valeriocos. I will update the tests for the PR in 1-2 days and then get it reviewed. 
Also the documentation for the tests seems really nice idea to me as it would make things quick and easier,  and I would be happy working on it!

**Comment by harshalmittal4** [comment]
@Polaris000 we can discuss regarding this shortly once we both have taken an overview :)

**Comment by harshalmittal4** [comment]
Hey @valeriocos, I have updated the PR.
Still the caching isn't very clear.

Caching is included currently but I am not sure about it. I saw #519 for that and found that caching isn't required in case of pull_request reviews (since no 2 PR's would have the same endpoints for reviews), but we need caching in this case, since same commit can be present in 2 different PR's (also if I remove the current caching part from the code, it fails for some repos). 

The test data for `github_request_pull_request_1_commits` and `github_request_pull_request_2_commits` used the same commit, which gave error (due to the caching that is used currently), so I needed to modify the commit_sha for pull_request_2.

Needed some thoughts regarding the caching implementation!
(Also sorry for the delay, this wouldn't be the case after my college exams are over by 4th May :) ).

**Comment by harshalmittal4** [comment]
Hey @valeriocos, I have updated the PR.
Still the caching isn't very clear.

Caching is included currently but I am not sure about it. I saw #519 for that and found that caching isn't required in case of pull_request reviews (since no 2 PR's would have the same endpoints for reviews), but we need caching in this case, since same commit can be present in 2 different PR's (also if I remove the current caching part from the code, it fails for some repos). 

The test data for `github_request_pull_request_1_commits` and `github_request_pull_request_2_commits` used the same commit, which gave error (due to the caching that is used currently), so I needed to modify the commit_sha for pull_request_2.

Needed some thoughts regarding the caching implementation!
(Also sorry for the delay, this wouldn't be the case after my college exams are over by 4th May ).

**Comment by valeriocos** [comment]
no worries @harshalmittal4 , let me know when I can review the PR, thanks! :)

**Comment by harshalmittal4** [comment]
The PR can be reviewed @valeriocos, but I needed help regarding the caching!

**Comment by harshalmittal4** [comment]
I will include the minor updates, once the caching part is reviewed.

**Comment by harshalmittal4** [comment]
Hey @valeriocos, the PR is good now, I was able to resolve the caching error, realised the problem!
Please have a look when you have time, thanks :)
(I would squash the commits once the review is done).

File: perceval/backends/core/github.py [hunk 4]
```diff
@@ -502,6 +522,7 @@ class GitHubClient(HttpClient, RateLimitHandler):
 
     _users = {}       # users cache
     _users_orgs = {}  # users orgs cache
+    _commit_comments = {}  # commit_comments cache
```
Comment (valeriocos) [inline perceval/backends/core/github.py]: it is probably better to call this variable `_commit hashes` or `_commits`, since it is used to store hashes. What do you think ?

File: tests/test_github.py [hunk 5]
```diff
@@ -731,6 +767,14 @@ def test_fetch_pulls_from_issues(self):
...
                                    'X-RateLimit-Reset': '15'
                                })
+        httpretty.register_uri(httpretty.GET,
+                               GITHUB_PULL_REQUEST_1_COMMIT_COMMENTS,
+                               body=pull_request_1_commit_comments,
+                               status=200,
+                               forcing_headers={
+                                   'X-RateLimit-Remaining': '20',
+                                   'X-RateLimit-Reset': '5'
+                               })
```
Comment (valeriocos) [inline tests/test_github.py]: In the test `test_fetch_pulls_from_issues` it seems that no assertion is done over the commit comments data

File: tests/test_github.py [hunk 6]
```diff
@@ -975,6 +1020,16 @@ def test_fetch_pulls_until_date(self):
                                    'Link': '<' + GITHUB_ISSUES_URL + '/?&page=2>; rel="next", <' +
                                            GITHUB_ISSUES_URL + '/?&page=3>; rel="last"'
                                })
+
+        httpretty.register_uri(httpretty.GET,
```
Comment (valeriocos) [inline tests/test_github.py]: In the test `test_fetch_pulls_until_date` it seems that no assertion is done over the commit comments data

File: tests/test_github.py [hunk 7]
```diff
@@ -1260,6 +1316,15 @@ def test_fetch_pull_enterprise(self):
                                    'Link': '<' + GITHUB_ENTERPRISE_ISSUES_URL + '/?&page=2>; rel="next", <' +
                                            GITHUB_ENTERPRISE_ISSUES_URL + '/?&page=2>; rel="last"'
                                })
+
+        httpretty.register_uri(httpretty.GET,
```
Comment (valeriocos) [inline tests/test_github.py]: In the test `test_fetch_pull_enterprise` it seems that no assertion is done over the commit comments data

**Review by valeriocos — COMMENTED** [review]
Thank you @harshalmittal4 for updating the PR. The code looks great, but there are some minor things to fix (see comments).

Please remember to update the github version number at: https://github.com/chaoss/grimoirelab-perceval/blob/master/perceval/backends/core/github.py#L88.

Thanks! :)

File: perceval/backends/core/github.py (hunk 4 above)
Comment (harshalmittal4) [reply perceval/backends/core/github.py]: Hey @valeriocos,
As the `_users` is storing the response from the endpoint `https://api.github.com/users/<login>` so that the same request is not issued again (if the same user comes),
similarly `_commit_comments` is storing the response from the endpoint `https://api.github.com/repos/<owner>/<repository>/commits/<commit-hash>/comments`, which gives the comments on that particular commit

File: tests/test_github.py (hunk 5 above)
Comment (harshalmittal4) [reply tests/test_github.py]: Sure, I'll add it and few more tests!

File: tests/test_github.py (hunk 6 above)
Comment (harshalmittal4) [reply tests/test_github.py]: Sure, will add it ;)

File: tests/test_github.py (hunk 7 above)
Comment (harshalmittal4) [reply tests/test_github.py]: Sure, will add it ;)

File: tests/test_github.py [hunk 8]
```diff
@@ -71,6 +71,11 @@
 GITHUB_ORGS_URL = GITHUB_API_URL + "/users/zhquan_example/orgs"
 GITHUB_COMMAND_URL = GITHUB_API_URL + "/command"
 
+GITHUB_PULL_REQUEST_1_COMMIT_COMMENTS = GITHUB_REPO_URL + \
```
Comment (harshalmittal4) [inline tests/test_github.py]: @valeriocos 
Also, one more thing which I needed to ask is since the mocked commits in `github_request_pull_request_1_commits` and `github_request_pull_request_2_commits` data are both the same ('53b970ee04bbc435842c14a2cbfdd623faf74a65'), shall I define the same url twice as done here, or is it good to define it only once `GITHUB_PULL_REQUEST_COMMIT_COMMENTS`( no `_1_` or `_2_`) ?
Another solution is to change the mocked commit in one of the pull_request to say '53b970ee04bbc435842c14a2cbfdd623faf74a6**6**', and then define these two seperate urls.

Comment (valeriocos) [reply tests/test_github.py]: I would go for your `Another solution`, thus (if I understood it correctly) we should have two test files, one file including the comments for a commit in PR 1 and another file including the comments for a commit in PR 2 (as done the test data of pull request comments or commits).  What do you think ?

**Comment by harshalmittal4** [comment]
@valeriocos, I have updated the PR and had some minor things to ask, please look into them, thanks :)

File: tests/test_github.py (hunk 8 above)
Comment (harshalmittal4) [reply tests/test_github.py]: Yes @valericos, this seemed better option to me!
The content for those two files would exactly be same, as I understand.

Comment (valeriocos) [reply tests/test_github.py]: Probably it's better to modify one of them. For instance, you can have one file with 4 comments and the other one with 2/3. The assertions in the tests could check the number of comments.
What do you think ?

**Review by valeriocos — COMMENTED** [review]
Thank you for the quick reaction @harshalmittal4  (a couple of comments left)

I was looking for a repo with commit comments to make a live test, but I haven't found any so far. I tried to add a comment to the PR https://github.com/chaoss/grimoirelab-graal/pull/26 in graal, but I didn't find how to do it (although github documentation says that is possible: https://developer.github.com/v3/guides/working-with-comments/). At the end I added two comments to graal (see below), which were correctly retrieved with your feature.
- https://github.com/chaoss/grimoirelab-graal/commit/ea891e1c48a9c26337045c0eb74329ec81475a9a
- https://github.com/chaoss/grimoirelab-graal/commit/e4dca9fbf827cdce17a29b5fd708e5d4fd179068

I'm a bit puzzled wrt commit comments as part of pull request reviews, since I wasn't able to add a comment to a PR commit, but the offical doc says that it's possible. So I would ask you to check it on your side to be sure that commit comments can be part of a pull request review. If this wasn't the case (due to some changes in the GitHub API), we can think about adding a new category to the github backend to retrieve commits and their comments (to discuss with @sduenas). What do you think ?

File: perceval/backends/core/github.py [hunk 9]
```diff
@@ -662,6 +683,23 @@ def pull_review_comments(self, pr_number):
...
+            return self._commit_comments[commit_sha]
+
+        payload = {
+            'per_page': PER_PAGE,
+            'direction': 'asc',
+            'sort': 'updated'
+        }
+        path = urijoin("commits", commit_sha, "comments")
+
+        items = [items for items in self.fetch_items(path, payload)]
```
[the remark this answers is not in the document]
Comment (harshalmittal4) [reply perceval/backends/core/github.py]: I tried it earlier and found that  `self.fetch_items(path, payload)` returns a generator object. Then the code `for items in self.fetch_items(path, payload)` gets a string which is a list of commit comments. So items[0] is just a string which comprises of the list of commit_comments.

File: tests/test_github.py (hunk 8 above)
Comment (harshalmittal4) [reply tests/test_github.py]: Okay, @valeriocos, would update it!

File: perceval/backends/core/github.py [hunk 10]
```diff
@@ -85,7 +85,7 @@ class GitHub(Backend):
     :param sleep_time: time to sleep in case
         of connection problems
     """
-    version = '0.20.1'
+    version = '0.20.2'
```
[the remark this answers is not in the document]
Comment (harshalmittal4) [reply perceval/backends/core/github.py]: Okay @valeriocos, will update it, sorry wasn't knowing about that.

File: perceval/backends/core/github.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (harshalmittal4) [reply perceval/backends/core/github.py]: I also was not able to find much for commit comments, so used this : [https://github.com/octocat/Spoon-Knife/pull/1176](https://github.com/octocat/Spoon-Knife/pull/1176) for the example.
The required pull_request (1176) data is quickly retrieved using the command`perceval github octocat Spoon-Knife --from-date "2019-03-06T23:25:42Z" --category pull_request --sleep-for-rate -t '3f7210425d61134182aa5c606079d1c63c651cee'>> testing.json 
`

[the remark this answers is not in the document]
Comment (valeriocos) [reply perceval/backends/core/github.py]: > I tried it earlier and found that self.fetch_items(path, payload) returns a generator object. Then the code for items in self.fetch_items(path, payload) gets a string which is a list of commit comments. So items[0] is just a string which comprises of the list of commit_comments.

If I'm not wrong the method `fetch_items` returns paginated items so `items[0]` should return just the first page (I'm going to check to confirm it)

> I also was not able to find much for commit comments, so used this : octocat/Spoon-Knife#1176 for the example...

Were you able to create one or more commit comments in a review ?

[the remark this answers is not in the document]
Comment (valeriocos) [reply perceval/backends/core/github.py]: I tried again to create a commit comment within a review but I didnt' find how. If we are not able to create one, it's possible that this feature is not supported anymore by GitHub (I'm really sorry about that @harshalmittal4). In this case, we should probably freeze the PR and contact GitHub, or see how we can reuse this code. What do you think @harshalmittal4 (any idea is welcomed)?

[the remark this answers is not in the document]
Comment (harshalmittal4) [reply perceval/backends/core/github.py]: hey @valeriocos, some views:
1) I am not able to figure out currently why the commit comments box at the end is showing up in some PR's but not in others, maybe github has discontinued that recently or it might be some other case (Since [this](https://github.com/yashkant/Padam-Tensorflow/pull/9/commits/095d5dcbccb4e1f28cc9b788c470bbc308d9179e) is an example of a random PR that I saw from Jan 2019, and there also the option for commit comments is not present at the end as it should have been).
image

2) I also wanted to ask regarding commenting on a specific line in the **Files changed**. Two options present : Add single line comment, and Start a review.
I suppose that when we select add single line comment, the comment is counted as a commit comment. Whereas any commit that is made after selecting the option Start a review (and before submitting the review) is counted as a review.
If this is the case, this PR can fetch the line comments (since they are counted as commit comments).

**Comment by valeriocos** [comment]
@harshalmittal4 sorry for the late reply!
WRT 1., would you like to contact GitHub and ask for clarifications ? I can do it if you want
WRT 2., it seems that in both cases these comments appear as review comments

**Comment by harshalmittal4** [comment]
Hey @valeriocos, 
w.r.t 1, I am OK with whatever you feel, no hard inclination to either side. Not having any problem if the PR is freezed for now as suggested earlier.
w.r.t 2, I will just update you after checking again! Thanks!

**Comment by harshalmittal4** [comment]
Hey @valeriocos, 
For the repo [https://github.com/octocat/Spoon-Knife/pull/1176](https://github.com/octocat/Spoon-Knife/pull/1176), the retrieved commit comments are present [here](https://jsonblob.com/fd34ac1c-6f38-11e9-83c8-096420647747), and they dont contain the line comment that I made, and also I am we are not able to figure out how the comments which are present for the three commits were made..so probably github might have removed support for such type of comments anymore.
(We can do whatever you feel would be good w.r.t to this PR since commit comments are no longer seen, just wondering if it is likely that commit comments will be enabled again by github :-P)

**Comment by harshalmittal4** [comment]
Shall I ask them on the Github API forum..?

**Comment by harshalmittal4** [comment]
I think I might just ask them regarding this on the Github API forum.

**Comment by Polaris000** [comment]
@harshalmittal4, I think it would be a great idea to ask. Nothing to lose (I think)!

**Comment by valeriocos** [comment]
Thank you for the feedback @harshalmittal4 . Yes, it would be good to ask (at least we know what's going on).

A possible idea to use your code could be to add another category for the github backend to retrieve commits. What do you think @sduenas @jgbarah ?

**Comment by harshalmittal4** [comment]
I just opened a topic regarding this feature in the Github API forum, lets see what comes up.

**Comment by valeriocos** [comment]
Great! thank you @harshalmittal4

**Comment by harshalmittal4** [comment]
@valeriocos, I got a reply to the post  [here](https://github.community/t5/GitHub-API-Development-and/feature-Commit-comments-for-a-Pull-request/m-p/23286#M1635), lets wait and see what comes or I'll ping once again. Thanks :)

**Comment by valeriocos** [comment]
Perfect! Thank you @harshalmittal4 for the heads-up

**Comment by harshalmittal4** [comment]
Hey @valeriocos 
As per the last reply obtained [here](https://github.community/t5/GitHub-API-Development-and/feature-Commit-comments-for-a-Pull-request/m-p/23286#M1635), commit comments are there but only as part of individual commits. They are not part of the commits present in a pull request. Let me know what you think about this!

**Comment by valeriocos** [comment]
Thank you for the info @harshalmittal4 ! I added a comment in the thread you linked.

We could add the code you wrote to support a new category (commits). However, Perceval already fetch git commits in a more efficient way using the git backend.

What do you think @sduenas , should we freeze the PR or try to reuse this code for another category ?

@jgbarah do you think it could be useful to have another category for the github backend to address some CHAOSS metrics ?

**Comment by sduenas** [comment]
See #515 with the reasons why I'm closing this PR.

**Comment by AnonKhamsuk** [comment]
> Addresses #515.
> 
> The changes made:
> 
> 
> 
> - `commits_data`  field in pull_request response originally contained the PR commit `hash`, now it contains `commit_comments` as well. 
> 
> 
> 
> - `commit_comments`  has two information : comment body and user_data (for the comment author).
> 
> 
> 
> Currently, the PR is not ready to be merged.
> 
> 
> 
> It is running into error `perceval.errors.ArchiveError: data storage error; cause: duplicated entry`, 
> 
> (the error can be reproduced using command  `perceval github Cloud-CV Origami --json-line --category pull_request --sleep-for-rate -t XXXX>> testing.json
> 
> ` , for any other other repo also, same error is popping up).
> 
> It would really be helpful if someone could have a look where am I missing!
> 
> 
> 
> Also, I wished to understand if any information needs to be removed or added in `commits_data`.
> 
> 
> 
> I will add the tests to the PR once this is done, thanks a lot :)
> 
> @valeriocos @aswanipranjal @jgbarah
